src/codegen.cpp

```diff
@@ -3654,7 +3654,10 @@ static jl_cgval_t emit_expr(jl_value_t *expr, jl_codectx_t *ctx, bool isboxed, b
     else {
         if (!strcmp(head->name, "$"))
             jl_error("syntax: prefix \"$\" in non-quoted expression");
-        if (jl_is_toplevel_only_expr(expr) &&
+        if ((jl_is_toplevel_only_expr(expr) ||
+                  head == abstracttype_sym  ||
+                  head == compositetype_sym ||
+                  head == bitstype_sym)     &&
```

Comment:
Maybe we should add these to `jl_is_toplevel_only_expr`. However I think there is a remaining problem if the expression uses local variables.